### PR TITLE
feat(ctcss): per-bookmark UI + threshold slider (#269 PR 3/3)

### LINKED ISSUE #269: Add CTCSS/DCS tone squelch
## Motivation

Standard feature on every scanner and commercial NFM radio. Lets users lock onto specific agencies, talkgroups, or repeaters on shared frequencies by decoding the sub-audible (67–254 Hz) tone or digital code that the transmitter sends alongside voice. Without it, scanning a shared frequency means hearing every user on the channel; with it, the squelch only opens for a specific tone/code match.

For the transcription use case, tone squelch is a huge quality multiplier — fewer false openings means fewer wasted inference runs and no cross-talk contamination in the transcript.

## Scope

- **CTCSS (analog tone)** — 42 standard sub-audible tones (67.0, 71.9, 74.4, …, 254.1 Hz). Goertzel-filter detection, selectable via UI.
- **DCS / CDCSS (digital code)** — 23-bit Golay code at 134.3 bit/s. Detected via matched filter on a high-pass-filtered baseband.
- **UI**: per-bookmark tone/code setting plus a global "Tone squelch mode" toggle (off / CTCSS / DCS). Dropdown for tone frequency / code value. Optional encode for TX support later (not a blocker for listen-only users).
- **DSP**: lives in sdr-radio's IF/AF chain, downstream of squelch and upstream of the AF output. Needs a high-pass to strip the sub-audible from speaker audio so users don't hear the tone, and a detector path that feeds the squelch gate.

## Non-goals

- P25 / DMR / NXDN / Motorola trunking decoders — different problem entirely, belongs to a separate \"digital voice modes\" epic
- CTCSS encode for TX — read-only SDR for now

## Design notes

- Goertzel filter is the standard cheap approach for CTCSS detection: single-frequency DFT tuned to each candidate tone, low CPU cost, well-documented. We'd run one Goertzel per selected tone (or all 42 in scan mode).
- DCS decoding is trickier — need a PLL-locked baseband + Golay decoder. Open-source references exist (dsd-fme, op25, libpj).
- Test signals: we can generate CTCSS test audio with GNURadio or pre-recorded samples from radioreference.com; no actual RF hardware needed for unit tests.

## Related

- Architectural context: we want to separate \"is there signal worth caring about\" (squelch/DSP layer) from \"what was said\" (transcription/ML layer) so improvements on one side flow downstream to the other. Tone squelch improves signal-worth-caring-about, which directly improves transcription accuracy in scanner use.
- Complements #265 (Auto Break / squelch-gated transcription) — with tone squelch, Auto Break only segments on transmissions from the specific agency/talkgroup the user cares about.

## Labels

\`enhancement\`, \`radio\`, \`dsp\`

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

Third and final PR for issue #269 (CTCSS sub-audible tone squelch). PR 1 (#285) shipped the pure-DSP Goertzel detector; PR 2 (#286) wired it into `af_chain` with a speaker-path HPF override and mode-switch persistence. This PR adds the user-facing controls, threshold slider, and bookmark round-trip.

- **CtcssMode serde** — `#[serde(tag = "kind", content = "hz")]` produces clean tagged JSON (`{"kind":"tone","hz":100.0}`).
- **Threshold backend** — new `AfChain::set_ctcss_threshold` / `RadioModule::set_ctcss_threshold` that rebuild the active detector and persist across mode switches. Threshold is applied BEFORE mode on the `set_mode` rebuild path so the constructed detector picks up the stored value.
- **Messages** — `UiToDsp::SetCtcssMode`, `UiToDsp::SetCtcssThreshold`, `DspToUi::CtcssSustainedChanged(bool)` (edge-triggered, only emitted on sustained-gate transitions).
- **RadioPanel UI** — three NFM-only rows: tone combo (Off + 51 tones), threshold spinner (0.05–0.50, default 0.10), read-only status row with three states (Off / Waiting / Detected).
- **State-leak fix** — leaving NFM force-clears the combo to Off so a user who enabled CTCSS on NFM can't end up stuck with muted audio and hidden controls after switching to WFM.
- **Bookmark schema** — `ctcss_mode` and `ctcss_threshold` as `Option<_>` with `#[serde(default)]`. Pre-PR-3 bookmarks deserialize to `None` which restore interprets as "leave current CTCSS alone".

## Out of scope / future

- DCS (digital code squelch) — separate future PR, needs PLL + Golay decoder.
- Auto-detect mode (scan all 51 tones) — not implemented; user picks the tone manually.
- Tone encode / TX — SDR is RX-only for now.

## Test plan

- [x] \`cargo test --workspace\` — 38 test suites pass, 0 failures
- [x] Triple-flavor clippy clean (whisper-cuda default, sherpa-cpu, sherpa-cuda)
- [x] \`cargo fmt --all -- --check\` clean
- [x] **Manual smoke test on whisper-cuda:**
  - [x] Visibility gating (CTCSS rows appear on NFM, hide on WFM/AM/SSB/CW)
  - [x] Combo selection — audio gates on tone mismatch, opens on match
  - [x] Status row label transitions: Off → Waiting → Detected
  - [x] Threshold slider — dragging changes detector sensitivity live
  - [x] Bookmark round-trip — save with CTCSS set, switch to Off, reload bookmark → CTCSS reapplies; app restart preserves it
  - [x] Pre-PR-3 bookmark backward compat — loads without touching current CTCSS
  - [x] Mode-switch state leak fix — NFM-with-tone → WFM returns audio; switching back to NFM leaves combo at Off
  - [x] HPF interaction — user HPF preference survives CTCSS engage/disengage cycles

Closes #269.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Full CTCSS (tone squelch) support in NFM: select from 52 tones, adjust detection threshold, and view sustained-gate status in the radio panel.
  * CTCSS mode and threshold are saved to and restored with bookmarks; UI applies threshold before mode on restore.

* **Bug Fixes**
  * Gate status updates are edge-triggered to avoid redundant notifications.
  * Threshold inputs are validated; invalid updates are rejected and reported.

### CODE REVIEW

File: crates/sdr-core/src/messages.rs [hunk 1]
```diff
@@ -37,6 +37,11 @@ pub enum DspToUi {
     /// transcription session (band change = new session boundary) and
     /// to re-run Auto Break row visibility rules.
     DemodModeChanged(DemodMode),
+    /// CTCSS sustained-gate state changed. Emitted only on edges
+    /// (closed → open / open → closed), not per-window, so the UI
+    /// status indicator can subscribe without flooding the channel.
+    /// Always `false` when CTCSS is currently `Off`.
+    CtcssSustainedChanged(bool),
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-core/src/messages.rs:44]: Good call — the other DspToUi/UiToDsp variants all have smoke tests so the new ones should match the pattern. Fixed in 0a32c62.

Three additions to `crates/sdr-core/src/messages.rs`:

1. **`ctcss_sustained_changed_message_constructs`** — new standalone test mirroring the existing `demod_mode_changed_message_constructs`. Constructs both the `true` and `false` variants and asserts the matches pattern works. Using a standalone test (not appending to `test_dsp_to_ui_variants`) because the existing pattern is inconsistent — `DemodModeChanged` lives in its own function and the older FFT/SignalLevel variants are bundled. Matching the newer pattern seemed cleaner for a message variant added in isolation.

2. **`SetCtcssMode(CtcssMode::Off)`** — appended to `test_ui_to_dsp_variants`, asserts the unit-variant matches.

3. **`SetCtcssMode(CtcssMode::Tone(100.0))` and `SetCtcssThreshold(0.15)`** — also appended to the same test, with `f32::EPSILON` comparisons matching the style of the adjacent `SetSquelch`/`SetNbLevel` float-carrying variants.

All three now instantiate cleanly and round-trip through the `matches!` pattern. The intent is exactly what you described — pin the protocol additions so a future rename/remove fails loudly at CI instead of silently compiling with the wrong signature.

File: crates/sdr-radio/src/lib.rs [hunk 2]
```diff
@@ -424,6 +435,26 @@ impl RadioModule {
...
+    pub fn set_ctcss_threshold(&mut self, threshold: f32) -> Result<(), RadioError> {
+        self.af_chain.set_ctcss_threshold(threshold)?;
+        self.ctcss_threshold = threshold;
+        Ok(())
+    }
+
+    /// Returns the current CTCSS detection threshold.
+    pub fn ctcss_threshold(&self) -> f32 {
+        self.ctcss_threshold
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/lib.rs:456]: Good request — threshold persistence + invalid rejection at the RadioModule boundary are exactly the kinds of contracts a future refactor could accidentally break. Fixed in 0a32c62.

Two new tests in `crates/sdr-radio/src/lib.rs` alongside the existing `test_radio_module_*` suite:

**`test_radio_module_ctcss_threshold_persists_across_set_mode`** — constructs a `RadioModule`, sets threshold to `0.25`, then asserts BOTH the cached value (`radio.ctcss_threshold()`) AND the AF chain value (`radio.af_chain().ctcss_threshold()`) report 0.25. Then switches modes NFM → AM and re-asserts both values. This exercises the round-trip the PR's set_mode re-apply path exists for: the AF chain is rebuilt from scratch on mode switch, so if the threshold cache on RadioModule isn't reapplied, the new AF chain snaps back to the DSP default (0.1) and silently un-tunes the user's setting. The test catches that specific regression.

Checking BOTH levels is deliberate — checking only `radio.ctcss_threshold()` would pass even if the AF chain reapply was broken (since the cache itself survives regardless). The invariant we actually care about is "what the detector will see when it next processes a window," which lives in af_chain, not on RadioModule.

**`test_radio_module_ctcss_threshold_rejects_invalid`** — sets a valid 0.2, then tries `0.0`, `-0.1`, `1.001`, `NaN`, `+Inf`, `-Inf`. All six must return `Err`. After all rejections, `radio.ctcss_threshold()` must still be `0.2` — the cache must NOT advance on failure. This pins the "store-after-validate" order in `set_ctcss_threshold`:

```rust
pub fn set_ctcss_threshold(&mut self, threshold: f32) -> Result<(), RadioError> {
    self.af_chain.set_ctcss_threshold(threshold)?;  // ← validates + may return Err
    self.ctcss_threshold = threshold;                 // ← only runs on Ok
    Ok(())
}
```

If someone flipped the order (cache first, validate second) the tests would catch it — after a rejection, the cached value would be whatever-was-rejected and the assertion `(0.2).abs() < 1e-6` would fail.

Both tests pass. Full workspace still green; triple-flavor clippy clean.

File: crates/sdr-radio/src/lib.rs [hunk 3]
```diff
@@ -654,4 +685,47 @@ mod tests {
...
+        assert!(radio.set_ctcss_threshold(0.0).is_err());
+        assert!(radio.set_ctcss_threshold(-0.1).is_err());
+        assert!(radio.set_ctcss_threshold(1.001).is_err());
+        assert!(radio.set_ctcss_threshold(f32::NAN).is_err());
+        assert!(radio.set_ctcss_threshold(f32::INFINITY).is_err());
+        assert!(radio.set_ctcss_threshold(f32::NEG_INFINITY).is_err());
+
+        // Cached value must still be the last successful one.
+        assert!((radio.ctcss_threshold() - 0.2).abs() < 1e-6);
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/lib.rs:797]: Fair — I caught myself in an inconsistency. My round-1 reply on the persistence test explicitly said "checking BOTH levels is deliberate — checking only `radio.ctcss_threshold()` would pass even if the AF chain reapply was broken," and I promptly forgot to apply the same discipline to the rejection test. The same argument applies in reverse: a rejection test that only checks the cache can miss a regression where the AF chain advances its internal state on an invalid value before the range check fires. Both layers need to be invariant under failure.

Fixed in 3dd3d56. Three changes to `test_radio_module_ctcss_threshold_rejects_invalid`:

1. **AF chain assertion added** — after every rejection, `radio.af_chain().ctcss_threshold()` must still be 0.2 alongside `radio.ctcss_threshold()`. Checking both levels pins the "store-after-validate" order at both the RadioModule wrapper AND the AF chain it delegates to.

2. **`matches!` with exact error variant** — took your suggested diff but made it a loop-iterated list so the test reads cleaner:

```rust
let invalid = [0.0_f32, -0.1, 1.001, f32::NAN, f32::INFINITY, f32::NEG_INFINITY];
for v in invalid {
    assert!(
        matches!(
            radio.set_ctcss_threshold(v),
            Err(RadioError::Dsp(DspError::InvalidParameter(_)))
        ),
        "threshold {v} should produce Err(RadioError::Dsp(DspError::InvalidParameter(_)))"
    );
    // ...
}
```

The error-variant match is important because a future refactor could wrap invalid-parameter errors in `RadioError::ModeSwitchFailed(String)` or some other variant and keep returning `Err(_)` — the old `is_err()` assertions would silently pass while losing the error-type contract. The tight `matches!` catches that.

3. **Bonus: re-assert invariants INSIDE the loop, not after.** This is a subtle point that your suggested diff didn't cover but I think matters: a post-loop single assertion only observes the final state. If rejection of value A corrupts layer X while rejection of value B corrupts layer Y, a post-loop assertion can miss the interleaved drift because layer X is re-corrected by some subsequent state or whatever. Loop-iterated assertions catch each corruption the moment it would happen.

All 4 CTCSS-related RadioModule tests pass. Triple-flavor clippy clean, fmt clean.

File: crates/sdr-radio/src/lib.rs [hunk 4]
```diff
@@ -654,4 +685,81 @@ mod tests {
...
+            assert!(
+                (radio.ctcss_threshold() - 0.2).abs() < 1e-6,
+                "RadioModule cache drifted after rejected value {v}"
+            );
+            assert!(
+                (radio.af_chain().ctcss_threshold() - 0.2).abs() < 1e-6,
+                "AF chain effective threshold drifted after rejected value {v}"
+            );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/lib.rs:797]: Fair — CLAUDE.md pins "named constants for all magic numbers (sample rates, buffer sizes, thresholds)" on non-test code, and you're right to apply the same discipline to test fixtures. Three places repeating `0.25` / `0.2` / `1e-6` is exactly the kind of thing that drifts silently across a future threshold-range tuning pass.

Fixed in 3a5c8fc. Took your suggested diff nearly verbatim but with one enhancement: each constant has a one-paragraph docstring explaining **why** the value was chosen, not just what it is. Future-me (or anyone else doing a tuning pass) gets the intent alongside the literal.

New test-module constants:

```rust
/// Float tolerance for CTCSS threshold round-trip equality.
/// `1e-6` comfortably exceeds f32 rounding error for the
/// single-assignment round-trips the tests exercise.
const CTCSS_TEST_EPS: f32 = 1e-6;

/// Non-default value used by the persistence test. Chosen
/// strictly inside the DSP-layer `(0, 1]` range and clearly
/// different from the `CTCSS_DEFAULT_THRESHOLD` (0.1) so a
/// regression that silently reverts to the default fails
/// loudly.
const CTCSS_PERSIST_THRESHOLD: f32 = 0.25;

/// "Last-good" baseline used by the rejection test. Any
/// in-range value would work; 0.2 is distinct from both the
/// DSP default (0.1) and the persistence test's 0.25 so
/// cross-test contamination would be noticeable.
const CTCSS_LAST_GOOD_THRESHOLD: f32 = 0.2;

/// Values that `set_ctcss_threshold` must reject. Covers the
/// boundary cases (0.0, just over 1.0), a sub-zero, and all
/// three non-finite IEEE-754 values. Used by
/// `test_radio_module_ctcss_threshold_rejects_invalid`.
const INVALID_CTCSS_THRESHOLDS: [f32; 6] =
    [0.0, -0.1, 1.001, f32::NAN, f32::INFINITY, f32::NEG_INFINITY];
```

The docstrings are slightly overkill for *test* constants, but they answer two questions a future reader will want to know:

1. **Why this specific value?** "0.25 because distinct from the 0.1 default" is a stronger signal than "0.25 because it fit". If someone later changes `CTCSS_DEFAULT_THRESHOLD` to 0.25, these tests become meaningless silent passes — the docstring is a breadcrumb to update the test fixture in the same pass.
2. **Why 1e-6 instead of f32::EPSILON?** Because the round-trip is single-assignment, not accumulated; `EPSILON` (~1.19e-7) would catch any regression `1e-6` catches, but `1e-6` is more forgiving for the "someone added a cast somewhere" refactor case and is how the other float tests in this crate spell it.

Both tests now use the constants throughout. All 4 CTCSS RadioModule tests pass; triple-flavor clippy clean; fmt clean.
